File: odbc-scenarios/duckdb-odbc-go-demo/internal/demo/duckdb_conn.go
// Single shared DuckDB connection, the Go analogue of DuckDbConfig.
//
// Why one connection? The ODBC extension stores its connection handle in a
// DuckDB *session variable* (SET VARIABLE conn = odbc_connect(...)), which is
// scoped to a single DuckDB connection. In the Java demo this forced a
// SingleConnectionDataSource instead of a pool. Here we use database/sql but
// pin the pool to exactly one underlying connection, so every statement runs
// on the same DuckDB session and SET VARIABLE state survives across requests.
// With ':memory:', each new connection would be its own database anyway, so
// sharing the one connection is essential, not just an optimization.
//
// database/sql already serializes callers onto a single connection when
// MaxOpenConns is 1, which keeps the single-connection model honest and matches
// the Java, Python, and Node demos.
package demo

import (
	"context"
	"database/sql"
	"fmt"

	duckdb "github.com/duckdb/duckdb-go/v2" // also registers the "duckdb" driver
)

// DuckDB wraps the single shared *sql.DB.
type DuckDB struct {
	db *sql.DB
}

// OpenDuckDB opens the in-memory (or file-backed) DuckDB and pins the pool to a
// single connection. Empty string opens an in-memory database.
func OpenDuckDB(database string) (*DuckDB, error) {
	// The duckdb driver treats an empty DSN as in-memory; translate our
	// ":memory:" convention (shared with the other demos) accordingly.
	dsn := database
	if dsn == ":memory:" {
		dsn = ""
	}

	db, err := sql.Open("duckdb", dsn)
	if err != nil {
		return nil, fmt.Errorf("open duckdb: %w", err)
	}

	// Pin to one physical connection so the ODBC session variable, loaded
	// extensions, and in-memory tables all persist across requests. Without
	// this, database/sql could open a second connection (fresh :memory: db) or
	// silently drop and recreate the idle one, losing all session state.
	db.SetMaxOpenConns(1)
	db.SetMaxIdleConns(1)
	db.SetConnMaxLifetime(0) // never expire
	db.SetConnMaxIdleTime(0) // never close while idle

	if err := db.Ping(); err != nil {
		return nil, fmt.Errorf("ping duckdb: %w", err)
	}
	return &DuckDB{db: db}, nil
}

// Close releases the underlying connection.
func (d *DuckDB) Close() error { return d.db.Close() }

// Exec runs a statement that returns no rows we care about (like
// JdbcTemplate.execute/update): INSTALL/LOAD, SET VARIABLE, ATTACH, COPY.
func (d *DuckDB) Exec(ctx context.Context, sql string, args ...any) error {
	_, err := d.db.ExecContext(ctx, sql, args...)
	return err
}

// Query runs a query and returns rows as plain JSON-safe maps (like
// JdbcTemplate.queryForList). Column order is not preserved (JSON objects are
// unordered), matching the other demos' object output.
func (d *DuckDB) Query(ctx context.Context, sql string, args ...any) ([]map[string]any, error) {
	rows, err := d.db.QueryContext(ctx, sql, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	cols, err := rows.Columns()
	if err != nil {
		return nil, err
	}

	var out []map[string]any
	for rows.Next() {
		values := make([]any, len(cols))
		ptrs := make([]any, len(cols))
		for i := range values {
			ptrs[i] = &values[i]
		}
		if err := rows.Scan(ptrs...); err != nil {
			return nil, err
		}
		row := make(map[string]any, len(cols))
		for i, c := range cols {
			row[c] = jsonSafe(values[i])
		}
		out = append(out, row)
	}
	if out == nil {
		out = []map[string]any{}
	}
	return out, rows.Err()
}

// jsonSafe converts driver values that don't marshal cleanly to JSON. DuckDB's
// Go driver returns most numerics natively, but a few types need help:
//   - DECIMAL comes back as duckdb.Decimal{Width, Scale, *big.Int}, which would
//     serialize as an ugly struct; render it as a string, matching the Node and
//     Python demos (so `revenue` is a string like "125000.50").
//   - []byte (e.g. BLOB) becomes a string instead of base64.
func jsonSafe(v any) any {
	switch t := v.(type) {
	case duckdb.Decimal:
		return t.String()
	case []byte:
		return string(t)
	default:
		return v
	}
}

File: odbc-scenarios/duckdb-odbc-rust-demo/src/duckdb_conn.rs
//! Single shared DuckDB connection, the Rust analogue of DuckDbConfig.
//!
//! Why one connection? The ODBC extension stores its connection handle in a
//! DuckDB *session variable* (SET VARIABLE conn = odbc_connect(...)), which is
//! scoped to a single DuckDB connection. In the Java demo this forced a
//! SingleConnectionDataSource instead of a pool. Here it is a
//! `duckdb::Connection`, which *is* one connection, so we create it once and
//! share it. Each new connection to ':memory:' would be its own database, so
//! sharing the handle is essential, not just an optimization.
//!
//! `duckdb::Connection` is `Send` but not `Sync`, so it cannot be shared across
//! threads concurrently. We wrap it in `Arc<Mutex<..>>`: the mutex serializes
//! access to the one connection - which is also what keeps the SET VARIABLE
//! state from interleaving across concurrent requests - and matches the
//! single-connection model of the Java, Python, and Node demos. (An r2d2 pool
//! would hand out *independent* connections that don't share the session
//! variable, so a pool is the wrong tool here.)

use std::sync::{Arc, Mutex};

use duckdb::types::ValueRef;
use duckdb::{params_from_iter, Connection};
use serde_json::{Map, Value};

/// A JSON-safe row: column name -> value, matching the other demos' output.
pub type Row = Map<String, Value>;

/// Shared handle to the single DuckDB connection.
#[derive(Clone)]
pub struct DuckDb {
    conn: Arc<Mutex<Connection>>,
}

impl DuckDb {
    /// Open the in-memory (or file-backed) DuckDB. ":memory:" (the default) maps
    /// to an in-memory database, matching the other demos' convention.
    pub fn open(database: &str) -> duckdb::Result<Self> {
        let conn = if database == ":memory:" {
            Connection::open_in_memory()?
        } else {
            Connection::open(database)?
        };
        Ok(DuckDb {
            conn: Arc::new(Mutex::new(conn)),
        })
    }

    /// Run one or more `;`-separated statements that return no rows we care
    /// about (INSTALL/LOAD, ATTACH, COPY). No parameters.
    pub fn execute_batch(&self, sql: &str) -> duckdb::Result<()> {
        let conn = self.conn.lock().unwrap();
        conn.execute_batch(sql)
    }

    /// Run a single statement with bound (string) parameters, e.g.
    /// `SET VARIABLE conn = odbc_connect(?, ?, ?)`.
    pub fn execute(&self, sql: &str, params: &[&str]) -> duckdb::Result<()> {
        let conn = self.conn.lock().unwrap();
        conn.execute(sql, params_from_iter(params.iter()))?;
        Ok(())
    }

    /// Run remote DDL/side-effecting SQL through the `odbc_query` table function.
    /// `odbc_query` is a *table* function, so it must sit in a `FROM` clause and
    /// its result set must be drained to force execution - even for DDL that
    /// returns nothing meaningful.
    pub fn drain(&self, sql: &str, params: &[&str]) -> duckdb::Result<()> {
        let conn = self.conn.lock().unwrap();
        let mut stmt = conn.prepare(sql)?;
        let mut rows = stmt.query(params_from_iter(params.iter()))?;
        while rows.next()?.is_some() {}
        Ok(())
    }

    /// Run a query and return rows as JSON-safe maps (like
    /// JdbcTemplate.queryForList). Column types vary per query, so each cell is
    /// converted from its DuckDB runtime type into a serde_json::Value.
    pub fn query(&self, sql: &str, params: &[&str]) -> duckdb::Result<Vec<Row>> {
        let conn = self.conn.lock().unwrap();
        let mut stmt = conn.prepare(sql)?;

        let mut rows = stmt.query(params_from_iter(params.iter()))?;

        // Column names are only available once the statement has been executed,
        // so read them from the started result set rather than the prepared
        // statement (column_names() panics if called before execution).
        let col_names: Vec<String> = match rows.as_ref() {
            Some(s) => s.column_names(),
            None => Vec::new(),
        };

        let mut out = Vec::new();
        while let Some(row) = rows.next()? {
            let mut obj = Map::new();
            for (i, name) in col_names.iter().enumerate() {
                obj.insert(name.clone(), value_to_json(row.get_ref(i)?));
            }
            out.push(obj);
        }
        Ok(out)
    }
}

/// Convert a DuckDB value into a JSON-safe value. DECIMAL and HUGEINT are
/// rendered as strings so large/exact values survive JSON without precision
/// loss, matching the Node and Python demos (`revenue` comes back as a string).
fn value_to_json(v: ValueRef<'_>) -> Value {
    match v {
        ValueRef::Null => Value::Null,
        ValueRef::Boolean(b) => Value::from(b),
        ValueRef::TinyInt(n) => Value::from(n),
        ValueRef::SmallInt(n) => Value::from(n),
        ValueRef::Int(n) => Value::from(n),
        ValueRef::BigInt(n) => Value::from(n),
        ValueRef::HugeInt(n) => Value::from(n.to_string()),
        ValueRef::UTinyInt(n) => Value::from(n),
        ValueRef::USmallInt(n) => Value::from(n),
        ValueRef::UInt(n) => Value::from(n),
        ValueRef::UBigInt(n) => Value::from(n),
        ValueRef::Float(f) => Value::from(f),
        ValueRef::Double(f) => Value::from(f),
        ValueRef::Decimal(d) => Value::from(d.to_string()),
        ValueRef::Text(bytes) => Value::from(String::from_utf8_lossy(bytes).into_owned()),
        ValueRef::Blob(bytes) => Value::from(String::from_utf8_lossy(bytes).into_owned()),
        other => Value::from(format!("{other:?}")),
    }
}
